store/store.go

```diff
@@ -0,0 +1,212 @@
+package store
+
+//import "log"
+import "time"
+import "golang.org/x/crypto/blake2b"
+import "sync"
+import "fmt"
+
+type NodeID [32]byte
+type Key NodeID
+
+type item struct {
+	value     string
+	expire    time.Time
+	republish time.Time
+	origPub   NodeID
+}
+
+type localItem struct {
+	value     string
+	repubTime time.Time
+}
+
+type items struct {
+	sync.RWMutex
+	m map[Key]item
+}
+
+type localItems struct {
+	sync.RWMutex
+	m map[Key]localItem
+}
+
+type replicate struct {
+	sync.RWMutex
+	time time.Time
+}
+
+type Database struct {
+	items      items
+	localItems localItems
+	ch         chan (localItem)
+	replicate  replicate
+	tExpire    time.Duration
+	tReplicate time.Duration
+	tRepublish time.Duration
+}
+
+func NewDatabase(tExpire, tReplicate, tRepublish time.Duration) (*Database, chan localItem) {
+	db := new(Database)
+
+	db.tExpire = tExpire
+	db.tReplicate = tReplicate
+	db.tRepublish = tRepublish
+
+	db.items = items{m: make(map[Key]item)}
+	db.localItems = localItems{m: make(map[Key]localItem)}
+	db.ch = make(chan localItem)
+
+	go db.itemHandler()
+	go db.republishHandler()
+
+	return db, db.ch
+}
+
+func (db *Database) setReplicate() {
+	db.replicate.Lock()
+	db.replicate.time = time.Now().Add(time.Duration(db.tReplicate))
+	db.replicate.Unlock()
+}
+
+func (db *Database) getReplicate() time.Time {
+	db.replicate.RLock()
+	time := db.replicate.time
+	db.replicate.RUnlock()
+
+	return time
+}
+
+// TODO: Max 1000 chars. Truncate input string.
+func (db *Database) AddItem(value string, origPub NodeID) error {
+	t := time.Now()
+	expire := t.Add(db.tExpire)
+	republish := t.Add(db.tRepublish)
+
+	key := Key(blake2b.Sum256([]byte(value)))
+
+	newItem := item{}
+
+	newItem.value = value
+	newItem.expire = expire
+	newItem.republish = republish
+	newItem.origPub = origPub
+
+	// TODO: Handle duplicate keys.
```

Comment:
We decided to not handle this after the talk with Erik, right?

Reply:
Yeah, old artifact.